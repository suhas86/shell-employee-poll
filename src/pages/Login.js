import { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { handleLogin } from '../actions/authedUser';

const Login = ({ dispatch, user, error }) => {
  const [userId, setUserId] = useState('');
  const [password, setPassword] = useState('');
  const [errorMessage, setErrorMessage] = useState('');
  const navigate = useNavigate();

  useEffect(() => {
    if (user) {
      navigate('/');
    }
  }, [navigate, user]);

  useEffect(() => {
    setErrorMessage(error);
  }, [error]);

  const handleOnChange = (e) => {
    const { name, value } = e.target;
    if (name === 'userId') {
      setUserId(value);
    } else if (name === 'password') {
      setPassword(value);
    }
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (userId === '' || password === '') {
      setErrorMessage('Please enter an email and password.');
      return;
    }
    dispatch(handleLogin({ userId, password }));
  };
  return (
    <div className="login flex-center-container ">
      <h4>Enter your login id and password</h4>
      <form onSubmit={handleSubmit}>
        {errorMessage && (
          <p data-testid="error-message" className="error-message">
            {errorMessage}
          </p>
        )}
        <label>
          Login Id:
          <input
            name="userId"
            data-testid="userId"
            type="text"
            onChange={handleOnChange}
          />
        </label>
        <label>
          Password:
          <input
            name="password"
            type="password"
            data-testid="password"
            onChange={handleOnChange}
          />
        </label>
        <button type="submit" data-testid="login">
          Login
        </button>
      </form>
    </div>
  );
};
const mapStateToProps = ({ authedUser }) => ({
  user: authedUser.user,
  error: authedUser.error,
});
export default connect(mapStateToProps)(Login);
