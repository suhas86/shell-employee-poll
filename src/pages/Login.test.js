import { screen, fireEvent } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Login from './Login';
import { store } from '../index';
import renderWithRedux from '../utils/testing/renderWithRedux';

const mockStore = configureStore([thunk]);

describe('Login Page', () => {
  it('renders without crashing', () => {
    renderWithRedux(<Login />, { store });
    expect(
      screen.getByText('Enter your login id and password')
    ).toBeInTheDocument();
  });

  it('Should show error if login id and password are empty', () => {
    renderWithRedux(<Login />, { store });
    const submitButton = screen.getByTestId('login');
    fireEvent.click(submitButton);
    expect(
      screen.getByText('Please enter an email and password.')
    ).toBeInTheDocument();
  });

  it('Should show error if login id is empty', () => {
    renderWithRedux(<Login />, { store });
    const passwordField = screen.getByTestId('password');
    fireEvent.change(passwordField, { target: { value: 'password' } });
    const submitButton = screen.getByTestId('login');
    fireEvent.click(submitButton);
    expect(
      screen.getByText('Please enter an email and password.')
    ).toBeInTheDocument();
  });
  it('Should show error if password is empty', () => {
    renderWithRedux(<Login />, { store });
    const loginField = screen.getByTestId('userId');
    fireEvent.change(loginField, { target: { value: 'login' } });

    const submitButton = screen.getByTestId('login');
    fireEvent.click(submitButton);
    expect(
      screen.getByText('Please enter an email and password.')
    ).toBeInTheDocument();
  });
  it('Should show error if login id and password are wrong', async () => {
    const store = mockStore({
      authedUser: { user: null, error: 'Incorrect username or password' },
    });
    renderWithRedux(<Login />, { store });
    expect(screen.getByTestId('error-message')).toBeInTheDocument();
  });
});
