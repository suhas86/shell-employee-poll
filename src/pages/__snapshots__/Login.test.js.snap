// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page renders without crashing 1`] = `undefined`;
